src/mpid/ch4/netmod/ofi/ofi_rma.h

```diff
@@ -121,22 +123,47 @@ static inline int MPIDI_OFI_query_datatype(MPI_Datatype dt,
 
-    if (*count == MPIDI_OFI_QUERY_ATOMIC_COUNT)
-        *count = MPIDI_Global.win_op_table[dt_index][op_index].max_atomic_count;
-
-    if (*count == MPIDI_OFI_QUERY_FETCH_ATOMIC_COUNT)
-        *count = MPIDI_Global.win_op_table[dt_index][op_index].max_fetch_atomic_count;
-
-    if (*count == MPIDI_OFI_QUERY_COMPARE_ATOMIC_COUNT)
-        *count = MPIDI_Global.win_op_table[dt_index][op_index].max_compare_atomic_count;
-
-    if (((int) *fi_dt) == -1 || ((int) *fi_op) == -1)
-        rc = -1;
-    else
-        rc = MPI_SUCCESS;
+    switch (query_type) {
+        case MPIDI_OFI_QUERY_ATOMIC_COUNT:
+            *count = MPIDI_Global.win_op_table[dt_index][op_index].max_atomic_count;
+            break;
+        case MPIDI_OFI_QUERY_FETCH_ATOMIC_COUNT:
+            *count = MPIDI_Global.win_op_table[dt_index][op_index].max_fetch_atomic_count;
+            break;
+        case MPIDI_OFI_QUERY_COMPARE_ATOMIC_COUNT:
+            *count = MPIDI_Global.win_op_table[dt_index][op_index].max_compare_atomic_count;
+            break;
+        default:
+            MPIR_Assert(*count == MPIDI_OFI_QUERY_ATOMIC_COUNT ||
+                        *count == MPIDI_OFI_QUERY_FETCH_ATOMIC_COUNT ||
+                        *count == MPIDI_OFI_QUERY_COMPARE_ATOMIC_COUNT);
+            break;
+    }
 
-    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPIDI_OFI_QUERY_DATATYPE);
-    return rc;
+    /* If same_op_no_op is set, we also check other processes' atomic status.
+     * We have to always fallback to AM unless all possible reduce and cswap
+     * operations are supported by provider. This is because the noop process
+     * does not know what the same_op is on the other processes, which might be
+     * unsupported (e.g., maxloc). Thus, the noop process has to always use AM, and
+     * other processes have to also only use AM in order to ensure atomicity with
+     * atomic get. The user can use which_accumulate_ops hint to specify used reduce
+     * and cswap operations.
+     * We calculate the max count of all specified ops for each datatype at window
+     * creation or info set. dtypes_max_count[dt_index] is the provider allowed
+     * count for all possible ops. It is 0 if any of the op is not supported and
+     * is a positive value if all ops are supported.
+     * Current <datatype, op> can become hw atomic only when both local op and all
+     * possible remote op are atomic. */
+
+    /* We use the minimal max_count of local op and any possible remote op as the limit
+     * to validate dt_size in later check. We assume the limit should be symmetric on
+     * all processes as long as the hint correctly contains the local op.*/
+    MPIR_Assert(*count >= (size_t) MPIDI_OFI_WIN(win).acc_hint->dtypes_max_count[dt_index]);
+
+    if (MPIDI_CH4U_WIN(win, info_args).accumulate_ops == MPIDI_CH4I_ACCU_SAME_OP_NO_OP)
+        *count = (size_t) MPIDI_OFI_WIN(win).acc_hint->dtypes_max_count[dt_index];
```

Comment:
Okay, I think now I finally started to understand the high-level logic here. If `same_op_no_op` and `which_accumulate_ops` includes non-HW-supported one (e.g. MAXLOC), then `dtypes_max_count` becomes zero and that will prevent the use of HW atomics. Correct?

Assuming I'm correct above, here are my summary of the data structures:
* `MPIDI_Global.win_op_table`: records OFI-level limits (which depends on the OFI provider)
* `acc_hint->dtypes_max_count`: records the actually usable atomic limits considering MPI-level restrictions (which depends on MPI standard and runtime infokeys e.g. `which_accumulate_ops`)

My confusion was mainly due to the fact that I did not recognize these two key structures and differences.

Reply:
Yes, your understanding is correct.

Let me add comment for each structure.

Reply:
Comment added

Reply:
I realized that you mostly explained the above ideas in the original comments. Apparently I missed to capture those.